File: BinaryNotesMQ/java/src/org/bn/mq/impl/InMemoryQueueStorage.java
/*
 Copyright 2006-2011 Abdulla Abdurakhmanov (abdulla@latestbit.com)
 Original sources are available at www.latestbit.com

 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at

 http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 */

package org.bn.mq.impl;

import java.util.HashMap;
import java.util.Iterator;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;

import org.bn.mq.IConsumer;
import org.bn.mq.IMessage;
import org.bn.mq.IMessageQueue;
import org.bn.mq.IPersistenceQueueStorage;
import org.bn.mq.IRemoteConsumer;

public class InMemoryQueueStorage<T> implements IPersistenceQueueStorage<T> {
    private String queueStorageName;
    private Map < String, List < IMessage<T> > > storage = new HashMap < String, List < IMessage<T> > > ();
    private final List < IMessage<T> > nullList = new LinkedList < IMessage<T> >();
    
    public InMemoryQueueStorage(String queueStorageName) {
        this.queueStorageName = queueStorageName;        
    }
    
    public List< IMessage<T> > getMessagesToSend(IConsumer<T> consumer) {
        List< IMessage<T> > result;
        synchronized(storage) {
            result = storage.get(consumer.getId());
            if(result==null)
                result=nullList;
        }    
        return result;
    }
    
    public void removeDeliveredMessage(String consumerId, String messageId) {
        synchronized(storage) {
            List< IMessage<T> > messages = storage.get(consumerId);
            if(messages!=null) {
                Iterator<IMessage<T>> iterator =  messages.iterator();
                for(IMessage<T> item: messages) {
                    if(item.getId().equals(messageId)) {
                        messages.remove(item);
                        break;
                    }
                }
            }
        }        
    }

    public void persistenceSubscribe(IConsumer<T> consumer) {
        synchronized(storage) {
            if(!storage.containsKey(consumer.getId())) {
                List < IMessage<T> > list = new LinkedList< IMessage<T> >();
                storage.put(consumer.getId(),list);
            }
        }
    }

    public void persistenceUnsubscribe(IConsumer<T> consumer) {
        synchronized(storage) {
            storage.remove(consumer.getId());
        }    
    }

    public void registerPersistenceMessage(IMessage<T> message) {
        synchronized(storage) {
            for( Map.Entry< String, List < IMessage<T> > > item : storage.entrySet() ) {
                item.getValue().add(message);
            }
        }    
    }

    public void close() {
    }
}

File: BinaryNotesMQ/.net/BinaryNotesMQ/src/org/bn/mq/impl/InMemoryQueueStorage.cs
/*
 Copyright 2006-2011 Abdulla Abdurakhmanov (abdulla@latestbit.com)
 Original sources are available at www.latestbit.com

 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at

 http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 */
using System;
using System.Collections.Generic;
using org.bn.mq.protocol;
using org.bn.mq.net;

namespace org.bn.mq.impl
{
	
	public class InMemoryQueueStorage<T> : IPersistenceQueueStorage<T>
	{
		private string queueStorageName;
		private IDictionary<string, IList<IMessage<T>>> storage = new Dictionary<string, IList<IMessage<T>>>();

        private IList<IMessage<T>> nullList = new List<IMessage<T>>();
		
		public InMemoryQueueStorage(string queueStorageName)
		{
			this.queueStorageName = queueStorageName;
		}

        public virtual IList<IMessage<T>> getMessagesToSend(IConsumer<T> consumer)
		{
            IList<IMessage<T>> result = nullList;
			lock (storage)
			{
                if (storage.ContainsKey(consumer.Id))
                {
                    result = storage[consumer.Id];
                }
			}
			return result;
		}

        public void removeDeliveredMessage(string consumerId, string messageId)
		{
			lock (storage)
			{
                if (storage.ContainsKey(consumerId))
                {
				    IList<IMessage<T>> messages = storage[consumerId];
					foreach(IMessage<T> item in messages)
					{
						if (item.Id.Equals(messageId))
						{
							messages.Remove(item);
							break;
						}
					}
				}
			}
		}

        public virtual void persistenceSubscribe(IConsumer<T> consumer)
		{
			lock (storage)
			{
				if (!storage.ContainsKey(consumer.Id))
				{
                    IList<IMessage<T>> messages = new List<IMessage<T>>();
                    storage[consumer.Id] = messages;
				}
			}
		}

        public virtual void persistenceUnsubscribe(IConsumer<T> consumer)
		{
			lock (storage)
			{
				storage.Remove(consumer.Id);
			}
		}

        public virtual void registerPersistenceMessage(IMessage<T> message)
		{
			lock (storage)
			{
				foreach(IList<IMessage<T>>item in storage.Values)
				{
					item.Add(message);
				}
			}
		}
		
		public virtual void  close()
		{
		}
	}
}